app/session.py

```diff
@@ -0,0 +1,90 @@
+import uuid
+
+import flask
+import redis
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+import itsdangerous
+from flask.sessions import SessionMixin, SessionInterface
+from werkzeug.datastructures import CallbackDict
+
+SESSION_PREFIX = "session"
+
+
+class ServerSession(CallbackDict, SessionMixin):
+    def __init__(self, initial=None, sid=None):
+        def on_update(self):
+            self.modified = True
+
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.modified = False
+
+
+class RedisSessionStore(SessionInterface):
+    def __init__(self, redis, app):
+        self._redis = redis
+        self._app = app
+
+    def _get_signer(self, app) -> itsdangerous.Signer:
+        return itsdangerous.Signer(
+            app.secret_key, salt="session", key_derivation="hmac"
+        )
+
+    def _get_key(self, sid: str) -> str:
+        return f"{SESSION_PREFIX}:{sid}"
+
+    def open_session(self, app: flask.Flask, request: flask.Request):
+        unverified_sid = request.cookies.get(app.session_cookie_name)
+        if not unverified_sid:
+            unverified_sid = str(uuid.uuid4())
+            return ServerSession(sid=unverified_sid)
+        signer = self._get_signer(app)
+        try:
+            sid_as_bytes = signer.unsign(unverified_sid)
+            sid = sid_as_bytes.decode()
+        except itsdangerous.BadSignature:
+            unverified_sid = str(uuid.uuid4())
+            return ServerSession(sid=unverified_sid)
+
+        val = self._redis.get(self._get_key(sid))
+        if val is not None:
+            try:
+                data = pickle.loads(val)
+                return ServerSession(data, sid=sid)
+            except:
+                pass
+        return ServerSession(sid=unverified_sid)
+
+    def save_session(
+        self, app: flask.Flask, session: ServerSession, response: flask.Response
+    ):
+        domain = self.get_cookie_domain(app)
+        path = self.get_cookie_path(app)
+        httponly = self.get_cookie_httponly(app)
+        secure = self.get_cookie_secure(app)
+        expires = self.get_expiration_time(app, session)
+        val = pickle.dumps(dict(session))
+        self._redis.setex(
+            name=self._get_key(session.sid),
+            value=val,
+            time=int(app.permanent_session_lifetime.total_seconds()),
```

Comment:
I wonder if we can store `app.permanent_session_lifetime.total_seconds()` at init time so it doesn't need to be computed each time?

Reply:
It's a timedelta method. Just makes an arithmetic sum and multiply. Should be fast enough.